File: tests/pages/attendance_page.py
"""
Attendance Workspace Page Object (/attendance).
"""

from playwright.sync_api import Page, expect
from tests.pages.base_page import BasePage
from tests.components.punch_button import PunchButtonComponent
from tests.components.datatable import DataTableComponent
from tests.components.modal import ModalComponent


class AttendancePage(BasePage):
    PATH = "/attendance"

    def __init__(self, page: Page):
        super().__init__(page)
        self.punch_button = PunchButtonComponent(page)
        self.table = DataTableComponent(page, name="attendance")
        self.regularize_button = (
            page.locator("[data-testid='open-correction-modal-btn'], button")
            .filter(has_text="Submit Correction")
            .or_(page.locator("button").filter(has_text="Regularize"))
            .first
        )
        self.modal = ModalComponent(page)

        # Regularization form inputs (lazy locators)
        self.reg_check_in_input = page.locator("input[type='time']").first
        self.reg_check_out_input = page.locator("input[type='time']").nth(1)
        self.reg_reason_input = page.locator("textarea, input[placeholder*='reason']").first
        self.reg_submit_button = (
            page.locator("button[type='submit']")
            .or_(page.locator("button").filter(has_text="Submit Correction"))
            .or_(page.locator("button").filter(has_text="Submit Request"))
            .first
        )

    def open_regularization_modal(self):
        """Opens the attendance regularization request modal if button is present."""
        if self.regularize_button.is_visible():
            self.regularize_button.click()
            expect(self.modal.dialog.or_(self.page.locator("[role='dialog']")).first).to_be_visible()

    def submit_regularization(self, date_str: str, check_in: str, check_out: str, reason: str):
        """Submits an attendance correction request."""
        if self.regularize_button.is_visible():
            self.open_regularization_modal()
            if self.reg_check_in_input.is_visible():
                self.reg_check_in_input.fill(check_in)
            if self.reg_check_out_input.is_visible():
                self.reg_check_out_input.fill(check_out)
            if self.reg_reason_input.is_visible():
                self.reg_reason_input.fill(reason)
            if self.reg_submit_button.is_visible():
                expect(self.reg_submit_button).to_be_enabled()
                self.reg_submit_button.click()

File: e2e/pages/AttendancePage.ts
import { Page, Locator, expect } from "@playwright/test";
import { BasePage } from "./BasePage";

export class AttendancePage extends BasePage {
  readonly punchInButton: Locator;
  readonly punchOutButton: Locator;
  readonly requestCorrectionButtons: Locator;
  readonly correctionInInput: Locator;
  readonly correctionOutInput: Locator;
  readonly correctionReasonInput: Locator;
  readonly submitCorrectionButton: Locator;
  readonly approveCorrectionButtons: Locator;
  readonly rejectCorrectionButtons: Locator;

  constructor(page: Page, baseURL?: string) {
    super(page, baseURL);
    this.punchInButton = page.locator('[data-testid="punch-in-btn"]');
    this.punchOutButton = page.locator('[data-testid="punch-out-btn"]');
    this.requestCorrectionButtons = page.locator('button:has-text("Request Correction"), button:has-text("Correct")');
    this.correctionInInput = page.locator('[data-testid="correction-in-input"], input#reqInInput');
    this.correctionOutInput = page.locator('[data-testid="correction-out-input"], input#reqOutInput');
    this.correctionReasonInput = page.locator('[data-testid="correction-reason-input"], textarea#reasonInput');
    this.submitCorrectionButton = page.locator('[data-testid="submit-correction-btn"], button[type="submit"]:has-text("Submit")');
    this.approveCorrectionButtons = page.locator('[data-testid="approve-correction-btn"]');
    this.rejectCorrectionButtons = page.locator('[data-testid="reject-correction-btn"]');
  }

  async goto(): Promise<void> {
    await this.navigate("/attendance");
  }

  async assertLoaded(): Promise<void> {
    await expect(this.page).toHaveURL(/.*\/attendance.*/);
    await expect(this.page.locator("body")).toContainText(/Attendance/i);
  }

  async punchIn(): Promise<void> {
    if (await this.punchInButton.isVisible()) {
      await this.punchInButton.click();
    }
  }

  async punchOut(): Promise<void> {
    if (await this.punchOutButton.isVisible()) {
      await this.punchOutButton.click();
    }
  }

  async openFirstCorrectionModal(): Promise<void> {
    const btn = this.requestCorrectionButtons.first();
    await expect(btn).toBeVisible();
    await btn.click();
  }

  async submitCorrection(checkIn: string = "09:00", checkOut: string = "18:00", reason: string = "Biometric reader sync issue"): Promise<void> {
    await this.openFirstCorrectionModal();
    await this.correctionInInput.fill(checkIn);
    await this.correctionOutInput.fill(checkOut);
    await this.correctionReasonInput.fill(reason);
    await this.submitCorrectionButton.click();
  }
}
